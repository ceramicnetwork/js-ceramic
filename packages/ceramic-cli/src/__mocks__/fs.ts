'use strict'
const fs = jest.genMockFromModule<any>('fs')
fs.promises = jest.requireActual('fs').promises

interface MockFile {
    name: string;
    birthtime: Date;
    message: string;
}

interface MockFs {
    [filePath: string]: MockFile;
}

let mockFs: MockFs = {}

class MockStream {
    filePath: string
    writeFlag: string

    constructor(filePath: string, writeFlag: string) {
        this.filePath = filePath
        this.writeFlag = writeFlag
    }

    write(message: string): void {
        const prevFile = mockFs[this.filePath] || {
            name: this.filePath,
            birthtime: new Date(),
            message: ''
        }

        if (this.writeFlag === 'w') {
            mockFs[this.filePath] = {
                ...prevFile,
                message
            }
        } else if (this.writeFlag === 'a') {
            mockFs[this.filePath] = {
                ...prevFile,
                message: prevFile.message.concat(message)
            }
        }
    }

    end(): void {
        // pass
    }
}

function __initMockFs(files: Array<MockFile>): void {
    mockFs = {}
    files.forEach((file) => {
        mockFs[file.name] = file
    })
}

function __getMockFs(): MockFs {
    return mockFs
}

function createWriteStream(filePath: string, options: { flags: string }): MockStream {
    return new MockStream(filePath, options.flags)
}

async function rename(filePath: string, nextFilePath: string): Promise<void> {
    const file = mockFs[filePath]
    delete mockFs[filePath]
    mockFs[nextFilePath] = file
}

async function stat(filePath: string): Promise<MockFile> {
    return mockFs[filePath]
}

fs.__initMockFs = __initMockFs
fs.__getMockFs = __getMockFs
fs.createWriteStream = createWriteStream
fs.promises.rename = rename
fs.promises.stat = stat

module.exports = fs
