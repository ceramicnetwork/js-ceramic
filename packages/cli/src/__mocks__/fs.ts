'use strict'
const fs = jest.genMockFromModule<any>('fs')
fs.promises = jest.requireActual('fs').promises

interface MockFile {
    name: string;
    birthtime: Date;
    message: string;
    size: number;
}

interface MockFs {
    [filePath: string]: MockFile;
}

let mockFs: MockFs = {}

class MockStream {
    filePath: string
    writeFlag: string

    constructor(filePath: string, writeFlag: string) {
        this.filePath = filePath
        this.writeFlag = writeFlag
    }

    write(message: string): void {
        const prevFile = mockFs[this.filePath] || {
            name: this.filePath,
            birthtime: new Date(),
            message: '',
            size: 0
        }

        if (this.writeFlag === 'w') {
            mockFs[this.filePath] = {
                ...prevFile,
                message
            }
        } else if (this.writeFlag === 'a') {
            mockFs[this.filePath] = {
                ...prevFile,
                message: prevFile.message.concat(message)
            }
        }
    }

    end(): void {
        // pass
    }
}

function __initMockFs(files: Array<MockFile>): void {
    mockFs = {}
    files.forEach((file) => {
        mockFs[file.name] = file
    })
}

function __getMockFs(): MockFs {
    return mockFs
}

function createWriteStream(filePath: string, options: { flags: string }): MockStream {
    return new MockStream(filePath, options.flags)
}

function writeFile(filePath: string, data: any, options: { flag: string }): Promise<void> {
    const prevFile = mockFs[filePath] || {
        name: filePath,
        birthtime: new Date(),
        message: '',
        size: 0
    }

    if (options.flag === 'w') {
        mockFs[filePath] = {
            ...prevFile,
            message: data
        }
    } else if (options.flag === 'a') {
        mockFs[filePath] = {
            ...prevFile,
            message: prevFile.message.concat(data)
        }
    }
}

async function rename(filePath: string, nextFilePath: string): Promise<void> {
    const file = mockFs[filePath]
    delete mockFs[filePath]
    mockFs[nextFilePath] = file
}

async function stat(filePath: string): Promise<MockFile> {
    return mockFs[filePath]
}

fs.__initMockFs = __initMockFs
fs.__getMockFs = __getMockFs
fs.createWriteStream = createWriteStream
fs.promises.rename = rename
fs.promises.stat = stat
fs.promises.writeFile = writeFile

module.exports = fs
