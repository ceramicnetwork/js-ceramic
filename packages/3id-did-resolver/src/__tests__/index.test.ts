import { jest } from '@jest/globals'
import { readFileSync } from 'node:fs'

const VECTORS_PATH = new URL('./vectors.json', import.meta.url)
const vectors = JSON.parse(readFileSync(VECTORS_PATH, 'utf-8'))

jest.unstable_mockModule('cross-fetch', () => {
  const mockedCalls = vectors['http-mock']
  const fetchFunc = jest.fn(async (url: string, opts: any = {}) => ({
    ok: true,
    json: async () => {
      const call = mockedCalls[url]
      if (opts.method === 'post') {
        const result = call.find(({ expectedBody }) => opts.body === JSON.stringify(expectedBody))
        if (result?.response) {
          return result.response
        } else {
          throw new Error('Could not find response for http body: ' + opts.body)
        }
      }
      return call.response
    },
  }))
  return {
    default: fetchFunc
  }
})

import { Resolver } from 'did-resolver'

const DID_LD_JSON = 'application/did+ld+json'
const v1 = '3IDv1'
const v1unanchored = '3IDv1-unanchored'
const v0 = '3IDv0'
const v0NoUpdates = '3IDv0-no-updates'

const toLdFormat = (result) => {
  const newResult = { ...result }
  newResult.didResolutionMetadata.contentType = DID_LD_JSON
  newResult.didDocument['@context'] = 'https://www.w3.org/ns/did/v1'
  return newResult
}

describe('3ID DID Resolver', () => {
  jest.setTimeout(20000)
  let ceramic

  beforeAll(async () => {
    const { CeramicClient } = await import('@ceramicnetwork/http-client')
    ceramic = new CeramicClient()
  })

  describe('3IDv1', () => {
    test('getResolver works correctly', async () => {
      const ThreeIdResolver = await import('../index.js')
      const threeIdResolver = ThreeIdResolver.getResolver(ceramic)
      expect(Object.keys(threeIdResolver)).toEqual(['3'])
    })

    test.each(vectors[v1].queries)('resolves 3id documents correctly %#', async (query) => {
      const ThreeIdResolver = await import('../index.js')
      const threeIdResolver = ThreeIdResolver.getResolver(ceramic)
      const resolver = new Resolver(threeIdResolver)
      const did = vectors[v1].did + query.params[0]
      expect(await resolver.resolve(did)).toEqual(query.result)
      if (query.params[1]) {
        const didVTime = vectors[v1].did + query.params[1]
        expect(await resolver.resolve(didVTime)).toEqual(query.result)
      }
      expect(await resolver.resolve(did, { accept: DID_LD_JSON })).toEqual(toLdFormat(query.result))
    })

    test.each(vectors[v1unanchored].queries)(
      'resolves unanchored 3id documents correctly %#',
      async (query) => {
        const ThreeIdResolver = await import('../index.js')
        const threeIdResolver = ThreeIdResolver.getResolver(ceramic)
        const resolver = new Resolver(threeIdResolver)
        const did = vectors[v1unanchored].did + query.params[0]
        expect(await resolver.resolve(did)).toEqual(query.result)
        expect(await resolver.resolve(did, { accept: DID_LD_JSON })).toEqual(
          toLdFormat(query.result)
        )
      }
    )

    test('no commit found', async () => {
      const ThreeIdResolver = await import('../index.js')
      const threeIdResolver = ThreeIdResolver.getResolver(ceramic)
      const resolver = new Resolver(threeIdResolver)
      const did =
        'did:3:kjzl6cwe1jw1490n1846tytf2fypoqb8aokqxl6n0lgyp1undfilllca3ztltho?version-id=0'
      await expect(resolver.resolve(did)).resolves.toEqual({
        didResolutionMetadata: {
          error: 'invalidDid',
          message:
            'Error: No resolution for commit k3y52l7qbv1fry84jsppsl3l75diunv3uzyrjiun7o8omq55y7ic9ljrbweqj5pmo',
        },
        didDocument: null,
        didDocumentMetadata: {},
      })
    })
  })

  describe('3IDv0', () => {
    test('resolves 3id with no updates', async () => {
      const ThreeIdResolver = await import('../index.js')
      const threeIdResolver = ThreeIdResolver.getResolver(ceramic)
      const resolver = new Resolver(threeIdResolver)
      const query = vectors[v0NoUpdates].query
      const did = vectors[v0NoUpdates].did + query.params[0]
      expect(await resolver.resolve(did)).toEqual(query.result)
    })

    test.each(vectors[v0].queries)('resolves 3id documents correctly %#', async (query) => {
      const ThreeIdResolver = await import('../index.js')
      const threeIdResolver = ThreeIdResolver.getResolver(ceramic)
      const resolver = new Resolver(threeIdResolver)
      const did = vectors[v0].did + query.params[0]
      expect(await resolver.resolve(did)).toEqual(query.result)
      if (query.params[1]) {
        const didVTime = vectors[v0].did + query.params[1]
        expect(await resolver.resolve(didVTime)).toEqual(query.result)
      }
      expect(await resolver.resolve(did, { accept: DID_LD_JSON })).toEqual(toLdFormat(query.result))
    })
  })
})
