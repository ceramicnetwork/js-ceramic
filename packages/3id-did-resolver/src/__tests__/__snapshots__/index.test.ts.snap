// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`3ID DID Resolver Legacy (v0) resolves 3id v0 1`] = `
Object {
  "@context": "https://w3id.org/did/v1",
  "authentication": Array [
    Object {
      "publicKey": "did:3:bafyreiffkeeq4wq2htejqla2is5ognligi4lvjhwrpqpl2kazjdoecmugi#g75q7J22wuWRyHb",
      "type": "Secp256k1SignatureAuthentication2018",
    },
  ],
  "id": "did:3:bafyreiffkeeq4wq2htejqla2is5ognligi4lvjhwrpqpl2kazjdoecmugi",
  "keyAgreement": Array [
    Object {
      "controller": "did:3:bafyreiffkeeq4wq2htejqla2is5ognligi4lvjhwrpqpl2kazjdoecmugi",
      "id": "did:3:bafyreiffkeeq4wq2htejqla2is5ognligi4lvjhwrpqpl2kazjdoecmugi#FRNNWcZDvDZcpVC",
      "publicKeyBase58": "qFYM1z9dJUC1K7rphEDVCj8U1YTzeFDVXtsjTa2uSiS",
      "type": "X25519KeyAgreementKey2019",
    },
  ],
  "publicKey": Array [
    Object {
      "controller": "did:3:bafyreiffkeeq4wq2htejqla2is5ognligi4lvjhwrpqpl2kazjdoecmugi",
      "id": "did:3:bafyreiffkeeq4wq2htejqla2is5ognligi4lvjhwrpqpl2kazjdoecmugi#g75q7J22wuWRyHb",
      "publicKeyBase58": "P8fYfHCRRHYkNwtVjRadhgftUsyEr6FUy65WttgSjqaW2eDXvJeXuCZRTdFM9JXc72eS5EDuu52fsdthsuWN7TS1",
      "type": "Secp256k1VerificationKey2018",
    },
    Object {
      "controller": "did:3:bafyreiffkeeq4wq2htejqla2is5ognligi4lvjhwrpqpl2kazjdoecmugi",
      "id": "did:3:bafyreiffkeeq4wq2htejqla2is5ognligi4lvjhwrpqpl2kazjdoecmugi#FRNNWcZDvDZcpVC",
      "publicKeyBase58": "qFYM1z9dJUC1K7rphEDVCj8U1YTzeFDVXtsjTa2uSiS",
      "type": "Curve25519EncryptionPublicKey",
    },
  ],
}
`;

exports[`3ID DID Resolver Legacy (v0) resolves 3id v0 which includs ceramic updates 1`] = `
Object {
  "@context": "https://w3id.org/did/v1",
  "authentication": Array [
    Object {
      "publicKey": "did:3:bafyreiffkeeq4wq2htejqla2is5ognligi4lvjhwrpqpl2kazjdoecmugi#8VBUiUTCeHbTeTV",
      "type": "Secp256k1SignatureAuthentication2018",
    },
  ],
  "id": "did:3:bafyreiffkeeq4wq2htejqla2is5ognligi4lvjhwrpqpl2kazjdoecmugi",
  "keyAgreement": Array [
    Object {
      "controller": "did:3:bafyreiffkeeq4wq2htejqla2is5ognligi4lvjhwrpqpl2kazjdoecmugi",
      "id": "did:3:bafyreiffkeeq4wq2htejqla2is5ognligi4lvjhwrpqpl2kazjdoecmugi#6F7kmzdodfvUCo9",
      "publicKeyBase58": "4jQRvHW8RfnRfdTrsxmExsn24z2vCV4xsoGxKB2Pkq2P",
      "type": "X25519KeyAgreementKey2019",
    },
  ],
  "publicKey": Array [
    Object {
      "controller": "did:3:bafyreiffkeeq4wq2htejqla2is5ognligi4lvjhwrpqpl2kazjdoecmugi",
      "id": "did:3:bafyreiffkeeq4wq2htejqla2is5ognligi4lvjhwrpqpl2kazjdoecmugi#8VBUiUTCeHbTeTV",
      "publicKeyBase58": "2Bv3fQSJiGJsbUUfUdiPsXXZGG5cXqaSeYnieuJcUFAbm",
      "type": "Secp256k1VerificationKey2018",
    },
    Object {
      "controller": "did:3:bafyreiffkeeq4wq2htejqla2is5ognligi4lvjhwrpqpl2kazjdoecmugi",
      "id": "did:3:bafyreiffkeeq4wq2htejqla2is5ognligi4lvjhwrpqpl2kazjdoecmugi#6F7kmzdodfvUCo9",
      "publicKeyBase58": "4jQRvHW8RfnRfdTrsxmExsn24z2vCV4xsoGxKB2Pkq2P",
      "type": "Curve25519EncryptionPublicKey",
    },
  ],
}
`;

exports[`3ID DID Resolver adds IDX root as service 1`] = `
Object {
  "@context": "https://w3id.org/did/v1",
  "authentication": Array [
    Object {
      "publicKey": "did:3:k2t6wyfsu4pg0t2n4j8ms3s33xsgqjhtto04mvq8w5a2v5xo48idyz38l7ydki#8VBUiUTCeHbTeTV",
      "type": "Secp256k1SignatureAuthentication2018",
    },
  ],
  "id": "did:3:k2t6wyfsu4pg0t2n4j8ms3s33xsgqjhtto04mvq8w5a2v5xo48idyz38l7ydki",
  "keyAgreement": Array [
    Object {
      "controller": "did:3:k2t6wyfsu4pg0t2n4j8ms3s33xsgqjhtto04mvq8w5a2v5xo48idyz38l7ydki",
      "id": "did:3:k2t6wyfsu4pg0t2n4j8ms3s33xsgqjhtto04mvq8w5a2v5xo48idyz38l7ydki#6F7kmzdodfvUCo9",
      "publicKeyBase58": "4jQRvHW8RfnRfdTrsxmExsn24z2vCV4xsoGxKB2Pkq2P",
      "type": "X25519KeyAgreementKey2019",
    },
  ],
  "publicKey": Array [
    Object {
      "controller": "did:3:k2t6wyfsu4pg0t2n4j8ms3s33xsgqjhtto04mvq8w5a2v5xo48idyz38l7ydki",
      "id": "did:3:k2t6wyfsu4pg0t2n4j8ms3s33xsgqjhtto04mvq8w5a2v5xo48idyz38l7ydki#8VBUiUTCeHbTeTV",
      "publicKeyBase58": "2Bv3fQSJiGJsbUUfUdiPsXXZGG5cXqaSeYnieuJcUFAbm",
      "type": "Secp256k1VerificationKey2018",
    },
    Object {
      "controller": "did:3:k2t6wyfsu4pg0t2n4j8ms3s33xsgqjhtto04mvq8w5a2v5xo48idyz38l7ydki",
      "id": "did:3:k2t6wyfsu4pg0t2n4j8ms3s33xsgqjhtto04mvq8w5a2v5xo48idyz38l7ydki#6F7kmzdodfvUCo9",
      "publicKeyBase58": "4jQRvHW8RfnRfdTrsxmExsn24z2vCV4xsoGxKB2Pkq2P",
      "type": "Curve25519EncryptionPublicKey",
    },
  ],
}
`;

exports[`3ID DID Resolver resolves 3id document correctly 1`] = `
Object {
  "@context": "https://w3id.org/did/v1",
  "authentication": Array [
    Object {
      "publicKey": "did:3:k2t6wyfsu4pg0t2n4j8ms3s33xsgqjhtto04mvq8w5a2v5xo48idyz38l7ydki#8VBUiUTCeHbTeTV",
      "type": "Secp256k1SignatureAuthentication2018",
    },
  ],
  "id": "did:3:k2t6wyfsu4pg0t2n4j8ms3s33xsgqjhtto04mvq8w5a2v5xo48idyz38l7ydki",
  "keyAgreement": Array [
    Object {
      "controller": "did:3:k2t6wyfsu4pg0t2n4j8ms3s33xsgqjhtto04mvq8w5a2v5xo48idyz38l7ydki",
      "id": "did:3:k2t6wyfsu4pg0t2n4j8ms3s33xsgqjhtto04mvq8w5a2v5xo48idyz38l7ydki#6F7kmzdodfvUCo9",
      "publicKeyBase58": "4jQRvHW8RfnRfdTrsxmExsn24z2vCV4xsoGxKB2Pkq2P",
      "type": "X25519KeyAgreementKey2019",
    },
  ],
  "publicKey": Array [
    Object {
      "controller": "did:3:k2t6wyfsu4pg0t2n4j8ms3s33xsgqjhtto04mvq8w5a2v5xo48idyz38l7ydki",
      "id": "did:3:k2t6wyfsu4pg0t2n4j8ms3s33xsgqjhtto04mvq8w5a2v5xo48idyz38l7ydki#8VBUiUTCeHbTeTV",
      "publicKeyBase58": "2Bv3fQSJiGJsbUUfUdiPsXXZGG5cXqaSeYnieuJcUFAbm",
      "type": "Secp256k1VerificationKey2018",
    },
    Object {
      "controller": "did:3:k2t6wyfsu4pg0t2n4j8ms3s33xsgqjhtto04mvq8w5a2v5xo48idyz38l7ydki",
      "id": "did:3:k2t6wyfsu4pg0t2n4j8ms3s33xsgqjhtto04mvq8w5a2v5xo48idyz38l7ydki#6F7kmzdodfvUCo9",
      "publicKeyBase58": "4jQRvHW8RfnRfdTrsxmExsn24z2vCV4xsoGxKB2Pkq2P",
      "type": "Curve25519EncryptionPublicKey",
    },
  ],
}
`;
